File: backend/app/schemas/emq_v2.py
# =============================================================================
# Stratum AI - EMQ v2 Schemas
# =============================================================================
"""
Pydantic models for EMQ (Event Measurement Quality) v2 API endpoints.
These schemas match the frontend types defined in api/emqV2.ts.

Every measured field is Optional and None means "not measured", not zero. The
service behind these endpoints used to substitute a default wherever a tenant
had no persisted signal health - a score of 75, an SVI of 15.3, $24,350 of
recovered revenue - and these types were what made that expressible: a
required ``score: float`` leaves no way to say "nothing was measured".
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


# =============================================================================
# Base Schema
# =============================================================================
class EMQBaseSchema(BaseModel):
    """Base schema with common configuration."""

    model_config = ConfigDict(
        from_attributes=True,
        str_strip_whitespace=True,
    )


# =============================================================================
# EMQ Score Schemas
# =============================================================================
class EmqDriver(EMQBaseSchema):
    """EMQ driver component."""

    name: str
    value: float
    weight: float
    status: Literal["good", "warning", "critical"]
    # None when the same component was not measured the day before: there is
    # no direction to report against a day that was never recorded.
    trend: Literal["up", "down", "flat"] | None = None


class EmqScoreResponse(EMQBaseSchema):
    """EMQ score response with drivers."""

    score: float | None = Field(
        default=None, ge=0, le=100, description="Overall EMQ score; None when not measured"
    )
    previousScore: float | None = Field(
        default=None, ge=0, le=100, description="Previous period score; None when not measured"
    )
    confidenceBand: Literal["reliable", "directional", "unsafe"] | None = None
    drivers: list[EmqDriver]
    lastUpdated: str = Field(..., description="ISO timestamp of last update")


# =============================================================================
# Confidence Schemas
# =============================================================================
class ConfidenceFactor(EMQBaseSchema):
    """Factor contributing to confidence score."""

    name: str
    contribution: float
    status: Literal["positive", "negative", "neutral"]


class ConfidenceThresholds(EMQBaseSchema):
    """Threshold values for confidence bands."""

    reliable: float
    directional: float


class ConfidenceDataResponse(EMQBaseSchema):
    """Confidence band details response."""

    band: Literal["reliable", "directional", "unsafe"] | None = None
    score: float | None = None
    thresholds: ConfidenceThresholds
    factors: list[ConfidenceFactor]


# =============================================================================
# Playbook Schemas
# =============================================================================
class PlaybookItemResponse(EMQBaseSchema):
    """Playbook item for EMQ fixes."""

    id: str
    title: str
    description: str
    priority: Literal["critical", "high", "medium", "low"]
    owner: str | None = None
    estimatedImpact: float = Field(..., description="Estimated EMQ score improvement")
    estimatedTime: str | None = None
    platform: str | None = None
    status: Literal["pending", "in_progress", "completed"]
    actionUrl: str | None = None


class PlaybookItemUpdate(EMQBaseSchema):
    """Playbook item update request."""

    status: Literal["pending", "in_progress", "completed"] | None = None
    owner: str | None = None


# =============================================================================
# Incident Schemas
# =============================================================================
class EmqIncidentResponse(EMQBaseSchema):
    """EMQ incident event."""

    id: str
    type: Literal["incident_opened", "incident_closed", "degradation", "recovery"]
    title: str
    description: str | None = None
    timestamp: str = Field(..., description="ISO timestamp")
    platform: str | None = None
    severity: Literal["critical", "high", "medium", "low"]
    recoveryHours: float | None = None
    emqImpact: float | None = None


# =============================================================================
# Impact Schemas
# =============================================================================
class ImpactBreakdown(EMQBaseSchema):
    """Per-platform impact breakdown."""

    platform: str
    actualRoas: float
    # Nothing models the counterfactual "ROAS under perfect attribution", so
    # these stay None rather than asserting a 15% improvement.
    estimatedRoas: float | None = None
    confidence: float
    revenueImpact: float | None = None


class EmqImpactResponse(EMQBaseSchema):
    """ROAS impact estimate response."""

    totalImpact: float | None = None
    currency: str = "USD"
    breakdown: list[ImpactBreakdown]


# =============================================================================
# Volatility Schemas
# =============================================================================
class VolatilityDataPoint(EMQBaseSchema):
    """Weekly volatility data point."""

    date: str
    value: float


class EmqVolatilityResponse(EMQBaseSchema):
    """Signal Volatility Index response."""

    svi: float | None = Field(default=None, description="Signal Volatility Index")
    # A direction needs two ends to compare; None when there is too little
    # history to state one.
    trend: Literal["increasing", "decreasing", "stable"] | None = None
    weeklyData: list[VolatilityDataPoint]


# =============================================================================
# Autopilot Schemas
# =============================================================================
class AutopilotStateResponse(EMQBaseSchema):
    """Autopilot state response."""

    mode: Literal["normal", "limited", "cuts_only", "frozen"]
    reason: str | None = None
    # fact_actions_queue carries no budget column, so this is None until a
    # source exists. It was the queued row count multiplied by a flat $5,000.
    budgetAtRisk: float | None = None
    allowedActions: list[str]
    restrictedActions: list[str]


class AutopilotModeUpdate(EMQBaseSchema):
    """Autopilot mode update request."""

    mode: Literal["normal", "limited", "cuts_only", "frozen"]
    reason: str | None = None


# =============================================================================
# Benchmark Schemas (Super Admin)
# =============================================================================
class EmqBenchmarkResponse(EMQBaseSchema):
    """Platform benchmark data."""

    platform: str
    p25: float | None = None
    p50: float | None = None
    p75: float | None = None
    # This endpoint is not tenant-scoped, so it cannot report a tenant's own
    # score or where that score falls in the distribution.
    tenantScore: float | None = None
    percentile: float | None = None


# =============================================================================
# Portfolio Schemas (Super Admin)
# =============================================================================
class TopIssue(EMQBaseSchema):
    """Top issue affecting tenants."""

    driver: str
    affectedTenants: int


class BandDistribution(EMQBaseSchema):
    """Distribution of tenants by confidence band."""

    reliable: int
    directional: int
    unsafe: int


class EmqPortfolioResponse(EMQBaseSchema):
    """Portfolio overview for super admin."""

    totalTenants: int
    byBand: BandDistribution
    atRiskBudget: float | None = None
    avgScore: float | None = None
    topIssues: list[TopIssue]

File: frontend/src/api/emqV2.ts
/**
 * Stratum AI - EMQ v2 Enhanced API
 *
 * Event Measurement Quality endpoints with Trust Layer integration
 *
 * The paths below are the ones `emq_v2.router` actually mounts. It is
 * registered with no prefix, so its routes are `/tenants/{id}/emq/<thing>` -
 * this client asked for `/emq/v2/tenants/{id}/<thing>`, which never existed.
 * Every hook here 404'd, in every view that used one, and what rendered was
 * whatever `??` fallback the view had written down.
 *
 * Every measured field is nullable and null means "not measured". The backend
 * used to substitute a default (score 75, SVI 15.3, $24,350 of recovered
 * revenue) wherever a tenant had no persisted signal health; it now returns
 * null, and these types are what let a view say so instead of guessing.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient, ApiResponse } from './client';

// Types
export interface EmqDriver {
  name: string;
  value: number;
  weight: number;
  status: 'good' | 'warning' | 'critical';
  /** Null when the same component was not measured the day before. */
  trend: 'up' | 'down' | 'flat' | null;
}

export interface EmqScore {
  /** Null when the tenant has no measured signal health. */
  score: number | null;
  previousScore: number | null;
  confidenceBand: 'reliable' | 'directional' | 'unsafe' | null;
  /** Only the components that were measured; empty when none were. */
  drivers: EmqDriver[];
  lastUpdated: string;
}

export interface ConfidenceData {
  band: 'reliable' | 'directional' | 'unsafe' | null;
  score: number | null;
  thresholds: {
    reliable: number;
    directional: number;
  };
  factors: {
    name: string;
    contribution: number;
    status: 'positive' | 'negative' | 'neutral';
  }[];
}

export interface PlaybookItem {
  id: string;
  title: string;
  description: string;
  priority: 'critical' | 'high' | 'medium' | 'low';
  owner: string | null;
  estimatedImpact: number;
  estimatedTime: string | null;
  platform: string | null;
  status: 'pending' | 'in_progress' | 'completed';
  actionUrl: string | null;
}

export interface EmqIncident {
  id: string;
  type: 'incident_opened' | 'incident_closed' | 'degradation' | 'recovery';
  title: string;
  description: string | null;
  timestamp: string;
  platform: string | null;
  severity: 'critical' | 'high' | 'medium' | 'low';
  recoveryHours: number | null;
  emqImpact: number | null;
}

export interface EmqImpact {
  totalImpact: number | null;
  currency: string;
  breakdown: {
    platform: string;
    actualRoas: number;
    /** Null: nothing models ROAS under perfect attribution. */
    estimatedRoas: number | null;
    confidence: number;
    revenueImpact: number | null;
  }[];
}

export interface EmqVolatility {
  svi: number | null; // Signal Volatility Index; null when not measured
  /** Null when there is too little history to state a direction. */
  trend: 'increasing' | 'decreasing' | 'stable' | null;
  weeklyData: {
    date: string;
    value: number;
  }[];
}

export interface EmqBenchmark {
  platform: string;
  p25: number | null;
  p50: number | null;
  p75: number | null;
  /** Null: the benchmarks endpoint is not tenant-scoped. */
  tenantScore: number | null;
  percentile: number | null;
}

export interface EmqPortfolio {
  totalTenants: number;
  byBand: {
    reliable: number;
    directional: number;
    unsafe: number;
  };
  atRiskBudget: number | null;
  avgScore: number | null;
  topIssues: {
    driver: string;
    affectedTenants: number;
  }[];
}

export type AutopilotMode = 'normal' | 'limited' | 'cuts_only' | 'frozen';

export interface AutopilotState {
  mode: AutopilotMode;
  reason: string | null;
  /** Null: fact_actions_queue carries no budget column. */
  budgetAtRisk: number | null;
  allowedActions: string[];
  restrictedActions: string[];
}

// API Functions
export const emqV2Api = {
  /**
   * Get current EMQ score for a tenant
   */
  getEmqScore: async (tenantId: number, date?: string): Promise<EmqScore> => {
    const params = date ? { date } : {};
    const response = await apiClient.get<ApiResponse<EmqScore>>(
      `/tenants/${tenantId}/emq/score`,
      { params }
    );
    return response.data.data;
  },

  /**
   * Get confidence band details
   */
  getConfidence: async (tenantId: number, date?: string): Promise<ConfidenceData> => {
    const params = date ? { date } : {};
    const response = await apiClient.get<ApiResponse<ConfidenceData>>(
      `/tenants/${tenantId}/emq/confidence`,
      { params }
    );
    return response.data.data;
  },

  /**
   * Get fix playbook items
   */
  getPlaybook: async (tenantId: number): Promise<PlaybookItem[]> => {
    const response = await apiClient.get<ApiResponse<PlaybookItem[]>>(
      `/tenants/${tenantId}/emq/playbook`
    );
    return response.data.data;
  },

  /**
   * Update playbook item status
   */
  updatePlaybookItem: async (
    tenantId: number,
    itemId: string,
    updates: Partial<Pick<PlaybookItem, 'status' | 'owner'>>
  ): Promise<PlaybookItem> => {
    const response = await apiClient.patch<ApiResponse<PlaybookItem>>(
      `/tenants/${tenantId}/emq/playbook/${itemId}`,
      updates
    );
    return response.data.data;
  },

  /**
   * Get incident timeline
   */
  getIncidents: async (
    tenantId: number,
    startDate: string,
    endDate: string
  ): Promise<EmqIncident[]> => {
    const response = await apiClient.get<ApiResponse<EmqIncident[]>>(
      `/tenants/${tenantId}/emq/incidents`,
      { params: { start_date: startDate, end_date: endDate } }
    );
    return response.data.data;
  },

  /**
   * Get ROAS impact estimate
   */
  getImpact: async (tenantId: number, startDate: string, endDate: string): Promise<EmqImpact> => {
    const response = await apiClient.get<ApiResponse<EmqImpact>>(
      `/tenants/${tenantId}/emq/impact`,
      { params: { start_date: startDate, end_date: endDate } }
    );
    return response.data.data;
  },

  /**
   * Get signal volatility data
   */
  getVolatility: async (tenantId: number, weeks?: number): Promise<EmqVolatility> => {
    const params = weeks ? { weeks } : {};
    const response = await apiClient.get<ApiResponse<EmqVolatility>>(
      `/tenants/${tenantId}/emq/volatility`,
      { params }
    );
    return response.data.data;
  },

  /**
   * Get autopilot state
   */
  getAutopilotState: async (tenantId: number): Promise<AutopilotState> => {
    const response = await apiClient.get<ApiResponse<AutopilotState>>(
      `/tenants/${tenantId}/emq/autopilot-state`
    );
    return response.data.data;
  },

  /**
   * Update autopilot mode (manual override)
   */
  updateAutopilotMode: async (
    tenantId: number,
    mode: AutopilotMode,
    reason?: string
  ): Promise<AutopilotState> => {
    const response = await apiClient.put<ApiResponse<AutopilotState>>(
      `/tenants/${tenantId}/emq/autopilot-mode`,
      { mode, reason }
    );
    return response.data.data;
  },

  // Super Admin endpoints
  /**
   * Get EMQ benchmarks across platform (super admin)
   */
  getBenchmarks: async (date?: string, platform?: string): Promise<EmqBenchmark[]> => {
    const params: Record<string, string> = {};
    if (date) params.date = date;
    if (platform) params.platform = platform;
    const response = await apiClient.get<ApiResponse<EmqBenchmark[]>>('/emq/benchmarks', {
      params,
    });
    return response.data.data;
  },

  /**
   * Get portfolio overview (super admin)
   */
  getPortfolio: async (date?: string): Promise<EmqPortfolio> => {
    const params = date ? { date } : {};
    const response = await apiClient.get<ApiResponse<EmqPortfolio>>('/emq/portfolio', {
      params,
    });
    return response.data.data;
  },
};

// React Query Hooks

export function useEmqScore(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['emq', 'score', tenantId, date],
    queryFn: () => emqV2Api.getEmqScore(tenantId, date),
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Refetch every 5 minutes
  });
}

export function useConfidence(tenantId: number, date?: string) {
  return useQuery({
    queryKey: ['emq', 'confidence', tenantId, date],
    queryFn: () => emqV2Api.getConfidence(tenantId, date),
    staleTime: 60 * 1000,
  });
}

export function useEmqPlaybook(tenantId: number) {
  return useQuery({
    queryKey: ['emq', 'playbook', tenantId],
    queryFn: () => emqV2Api.getPlaybook(tenantId),
    staleTime: 30 * 1000,
  });
}

export function useUpdatePlaybookItem(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      itemId,
      updates,
    }: {
      itemId: string;
      updates: Partial<Pick<PlaybookItem, 'status' | 'owner'>>;
    }) => emqV2Api.updatePlaybookItem(tenantId, itemId, updates),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['emq', 'playbook', tenantId] });
    },
  });
}

export function useEmqIncidents(tenantId: number, startDate: string, endDate: string) {
  return useQuery({
    queryKey: ['emq', 'incidents', tenantId, startDate, endDate],
    queryFn: () => emqV2Api.getIncidents(tenantId, startDate, endDate),
    staleTime: 60 * 1000,
  });
}

export function useEmqImpact(tenantId: number, startDate: string, endDate: string) {
  return useQuery({
    queryKey: ['emq', 'impact', tenantId, startDate, endDate],
    queryFn: () => emqV2Api.getImpact(tenantId, startDate, endDate),
    staleTime: 5 * 60 * 1000,
  });
}

export function useEmqVolatility(tenantId: number, weeks?: number) {
  return useQuery({
    queryKey: ['emq', 'volatility', tenantId, weeks],
    queryFn: () => emqV2Api.getVolatility(tenantId, weeks),
    staleTime: 60 * 1000,
  });
}

export function useAutopilotState(tenantId: number) {
  return useQuery({
    queryKey: ['emq', 'autopilot', tenantId],
    queryFn: () => emqV2Api.getAutopilotState(tenantId),
    staleTime: 30 * 1000,
    refetchInterval: 60 * 1000,
  });
}

export function useUpdateAutopilotMode(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ mode, reason }: { mode: AutopilotMode; reason?: string }) =>
      emqV2Api.updateAutopilotMode(tenantId, mode, reason),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['emq', 'autopilot', tenantId] });
    },
  });
}

// Super Admin hooks
export function useEmqBenchmarks(date?: string, platform?: string) {
  return useQuery({
    queryKey: ['emq', 'benchmarks', date, platform],
    queryFn: () => emqV2Api.getBenchmarks(date, platform),
    staleTime: 5 * 60 * 1000,
  });
}

export function useEmqPortfolio(date?: string) {
  return useQuery({
    queryKey: ['emq', 'portfolio', date],
    queryFn: () => emqV2Api.getPortfolio(date),
    staleTime: 5 * 60 * 1000,
  });
}
